File: jsrc/kernelprolog/fluents/SxxMachine/TermCollector.java
package SxxMachine;

import java.util.ArrayList;

/**
  Builds  Fluents from Java
  Streams
*/
public class TermCollector extends Sink {
	protected ArrayList buffer;

	private Prog p;

	public TermCollector(Prog p) {
		super(p);
		this.p = p;
		this.buffer = new ArrayList();
	}

	public int putElement(Term T) {
		buffer.add(T);
		return 1;
	}

	public void stop() {
		buffer = null;
	}

	public Term collect() {
		return new IterableSource(buffer, p);
	}
}

File: src/kernelprolog/fluents/SxxMachine/TermCollector.cpp
using namespace std;

#include "TermCollector.h"
#include "../../terms/SxxMachine/Prog.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../main/SxxMachine/Builtins.h"
#include "IterableSource.h"

namespace SxxMachine
{

	TermCollector::TermCollector(Prog *p) : Sink(p)
	{
		this->p = p;
		this->buffer = vector();
	}

	int TermCollector::putElement(Term *T)
	{
		buffer.push_back(T);
		return 1;
	}

	void TermCollector::stop()
	{
		buffer.clear();
	}

	Term *TermCollector::collect()
	{
		return new IterableSource(buffer, p);
	}
}
